File: 0973-k-closest-points-to-origin/0973-k-closest-points-to-origin.java
class Solution {
    public int[][] kClosest(int[][] points, int k) {
        // traverse
        // use heap to store points and their distances to the origin
        PriorityQueue<int[]> pq = new PriorityQueue<>((a,b)->Integer.compare(distance(b),distance(a)));
        
        for(int[] point: points){
            // add element into heap
            pq.offer(point);
            // remove redundant element
            if(pq.size() > k){
                pq.poll();
            }
        }
        int[][] result = new int[k][];

        // get top k elements from big heap:
        while(k > 0){
            result[k-1] = pq.poll();
            k--;
        }
        // return
        return result;
    }
    // helper function 
    public int distance(int[] i){
        return i[0]*i[0] + i[1]*i[1];
    }
}

File: 0973-k-closest-points-to-origin/0973-k-closest-points-to-origin.py
class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        # use max_heap to store points
        # use negative distance to simulate max_heap
        max_heap = [] # create list as heap
        for point in points:
            distance = -(point[0]*point[0] + point[1]*point[1]) 
            if len(max_heap) < k:
                heapq.heappush(max_heap, (distance,point))
            else:
                heapq.heappushpop(max_heap, (distance,point))
        return [point for (_, point) in max_heap]
        